packages/iov-bns/src/encode.ts

```diff
@@ -513,15 +515,15 @@ export function buildUnsignedTx(tx: UnsignedTransaction): codecImpl.bnsd.ITx {
       tx.fee && tx.fee.tokens
         ? {
             fees: encodeAmount(tx.fee.tokens),
-            payer: feePayer,
+            payer: feePayerAddressBytes,
           }
         : null,
     multisig: isMultisignatureTx(tx) ? tx.multisig.map(encodeNumericId) : null,
   });
 }
 
-export function buildSignedTx(tx: SignedTransaction): codecImpl.bnsd.ITx {
+export function buildSignedTx(tx: SignedTransaction, feePayer?: Address): codecImpl.bnsd.ITx {
```

Comment:
This signature seems to be incorrect. How can `tx` be "signed" when it does not contain the `feePayer` information?

I guess the `feePayer?` needs to become an attribute of `tx.transaction.fee`

Reply:
When the signature was created, the function `buildUnsignedTx` would have been called with the fee payer, meaning the fee payer is in the `bytesToSign` result. But by the time this gets back to `buildSignedTx` we've turned it back into a BCP tx object, so we've dropped that information and need to supply it again.

We can either make it part of the fee, or add a (potentially optional) `feePayer` attribute to a transaction and assume that only one person pays all the fees for the whole transaction.

Reply:
> by the time this gets back to buildSignedTx we've turned it back into a BCP tx object, so we've dropped that information and need to supply it again.

So it should be in the BCP tx object

> We can either make it part of the fee, or add a (potentially optional) feePayer attribute to a transaction

I think an optional `payer` in `Fee` is clean.

> and assume that only one person pays all the fees for the whole transaction

I am not aware of any case where multiple accounts could split the fee

Reply:
> assume that only one person pays all the fees for the whole transaction
I meant "assume that all the fees for the whole transaction are paid by only one person". The question is whether we individuate the fee payer on a per-transaction or per-fee basis. Per-fee is better aligned with the back end.

Reply:
> The question is whether we individuate the fee payer on a per-transaction or per-fee basis. Per-fee is better aligned with the back end.

I guess I don't understand. A sindle transaction has a single fee field (called plural "fees" for some reason). This single fee field can be unset or a single `FeeInfo` object, containing a single amount and a payer.

Reply:
Ah, sorry, my mistake, I was confusing the fees associated with different message types with the fee associated with a transaction.